pkg/operator/encryption/statemachine/transition.go

```diff
@@ -0,0 +1,238 @@
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/operatorclient"
+)
+
+func GetEncryptionConfigAndState(
+	podClient corev1client.PodsGetter,
+	secretClient corev1client.SecretsGetter,
+	targetNamespace string,
+	encryptionSecretSelector metav1.ListOptions,
+	encryptedGRs []schema.GroupResource,
+) (current *apiserverconfigv1.EncryptionConfiguration, desired map[schema.GroupResource]state.GroupResourceState, secretsFound bool, transitioningReason string, err error) {
+	revision, err := getAPIServerRevisionOfAllInstances(podClient.Pods(targetNamespace))
+	if err != nil {
+		return nil, nil, false, "", err
+	}
+	if len(revision) == 0 {
+		return nil, nil, false, "APIServerRevisionNotConverged", nil
+	}
+
+	encryptionConfig, err := getCurrentEncryptionConfig(secretClient.Secrets(targetNamespace), revision)
+	if err != nil {
+		return nil, nil, false, "", err
+	}
+
+	encryptionSecretList, err := secretClient.Secrets(operatorclient.GlobalMachineSpecifiedConfigNamespace).List(encryptionSecretSelector)
+	if err != nil {
+		return nil, nil, false, "", err
+	}
+	var encryptionSecrets []*corev1.Secret
+	for i, s := range encryptionSecretList.Items {
+		if _, err := secrets.ToKeyState(&s); err != nil {
+			klog.Infof("skipping invalid encryption secret %s", s.Name)
+			continue
+		} else {
+			encryptionSecrets = append(encryptionSecrets, &encryptionSecretList.Items[i])
+		}
+	}
+
+	desiredEncryptionState := getDesiredEncryptionState(encryptionConfig, encryptionSecrets, encryptedGRs)
+
+	return encryptionConfig, desiredEncryptionState, len(encryptionSecrets) > 0, "", nil
+}
+
+// getDesiredEncryptionState returns the desired state of encryption for all resources.
+// To do this it compares the current state against the available secrets and to-be-encrypted resources.
+// oldEncryptionConfig can be nil if there is no config yet.
+// If there are no secrets, the identity is set for all resources as write key.
+// It is assumed that encryptionSecrets are all valid.
+//
+// The basic rules are:
+//
+// 1. don't do anything if there are key secrets.
+// 2. every GR must have all the read-keys (existing as secrets) since last complete migration.
+// 3. if (2) is the case, the write-key must be the most recent key.
+// 4. if (2) and (3) are the case, all non-write keys should be removed.
+func getDesiredEncryptionState(oldEncryptionConfig *apiserverconfigv1.EncryptionConfiguration, encryptionSecrets []*corev1.Secret, toBeEncryptedGRs []schema.GroupResource) map[schema.GroupResource]state.GroupResourceState {
+	backedKeys := make([]state.KeyState, 0, len(encryptionSecrets))
+	for _, s := range encryptionSecrets {
+		km, err := secrets.ToKeyState(s)
+		if err != nil {
+			klog.Warningf("skipping invalid secret: %v", err)
+		}
+		backedKeys = append(backedKeys, km)
+	}
+	backedKeys = state.SortRecentFirst(backedKeys)
+
+	//
+	// STEP 0: start with old encryption config, and alter is in the rest of the rest of the func towards the desired state.
```

Comment:
this comment reads weird, could it be: `STEP 0: start with old encryption config, and alter it towards the desired state.` ?

Reply:
done